internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -0,0 +1,585 @@
+	// current (persisted) connection details, as well as any existing composed
+	// resource and their current connection details. The desired state includes
+	// only the XR and its connection details, which will initially be identical
+	// to the observed state.
+	xrConnDetails, err := c.composite.FetchConnection(ctx, xr)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errFetchXRConnectionDetails)
+	}
+	o, err := AsState(xr, xrConnDetails, observed)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errBuildObserved)
+	}
+	d, err := AsState(xr, xrConnDetails, ComposedResourceStates{})
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errBuildDesired)
+	}
+
+	events := []event.Event{}
+
+	// Run any Composition Functions in the pipeline. Each Function may mutate
+	// the desired state returned by the last, and each Function may produce
+	// results that will be emitted as events.
+	for _, fn := range req.Revision.Spec.Pipeline {
+		req := &v1beta1.RunFunctionRequest{Observed: o, Desired: d}
+
+		if fn.Input != nil {
+			in := &structpb.Struct{}
+			if err := in.UnmarshalJSON(fn.Input.Raw); err != nil {
+				return CompositionResult{}, errors.Wrapf(err, errFmtUnmarshalPipelineStepInput, fn.Step)
+			}
+			req.Input = in
+		}
+
+		// TODO(negz): Generate a content-addressable tag for this request.
+		// Perhaps using https://github.com/cerbos/protoc-gen-go-hashpb ?
+		rsp, err := c.pipeline.RunFunction(ctx, fn.FunctionRef.Name, req)
+		if err != nil {
+			return CompositionResult{}, errors.Wrapf(err, errFmtRunPipelineStep, fn.Step)
+		}
+
+		d = rsp.GetDesired()
+
+		// Results of fatal severity stop the Composition process. Other results
+		// are accumulated to be emitted as events by the Reconciler.
+		for _, rs := range rsp.Results {
+			switch rs.Severity {
+			case v1beta1.Severity_SEVERITY_FATAL:
+				return CompositionResult{}, errors.Errorf(errFmtFatalResult, fn.Step, rs.Message)
+			case v1beta1.Severity_SEVERITY_WARNING:
+				events = append(events, event.Warning(reasonCompose, errors.Errorf("Pipeline step %q: %s", fn.Step, rs.Message)))
+			case v1beta1.Severity_SEVERITY_NORMAL:
+				events = append(events, event.Normal(reasonCompose, fmt.Sprintf("Pipeline step %q: %s", fn.Step, rs.Message)))
+			case v1beta1.Severity_SEVERITY_UNSPECIFIED:
+				// We could hit this case if a Function was built against a newer
+				// protobuf than this build of Crossplane, and the new protobuf
+				// introduced a severity that we don't know about.
+				events = append(events, event.Warning(reasonCompose, errors.Errorf("Pipeline step %q returned a result of unknown severity (assuming warning): %s", fn.Step, rs.Message)))
+			}
+		}
+	}
+
+	// TODO(negz): Both here and in the PTComposer we're relying on the fact
+	// that the calling Reconciler will update xr's status in order to persist
+	// any changes we make to it. It would be more obvious if we owned making
+	// our own changes.
```

Comment:
+1, it took me a while to understand it, and it's clearly the cause of https://github.com/crossplane/crossplane/issues/3812, as you are suggesting below, we create a new xr one, but the caller is relying on the fact that we modify the resource in place. We should probably either move in here the status patching or explicitly return the xr as part of the CompositionResult

Reply:
I've added a commit that fixes this. Per https://github.com/crossplane/crossplane/issues/4581 we now _only_ support updating XR status. For now I've only made this change in the `FunctionComposer` to avoid extra scope creep in this PR. 😄